### PR TITLE
Sparsity Decay Schedules and Tests

### PR DESCRIPTION
**Pull request by xwinxu**

### CODE REVIEW

File: tensorflow_model_optimization/python/core/sparsity_tf2/schedule.py [hunk 1]
```diff
@@ -235,32 +278,54 @@ def __call__(self, step):
...
-               initial_lr=0.01,
                end_step=-1,
-               frequency=100):
+               frequency=100,
+               optimizer=None):
     """Initializes an annealing schedule for the drop fraction in a
     mask update function that follows the learning rate.
-
     Args:
       optimizer: the optimizer being used at training time
```
Comment (tomerk) [inline tensorflow_model_optimization/python/core/sparsity_tf2/schedule.py]: We talked about this offline: this should take a learning rate schedule instead of an optimizer to prevent subtle serialization/deserialization bugs in practice

File: tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py [hunk 2]
```diff
@@ -14,30 +14,361 @@
...
-from tensorflow_model_optimization.python.core.sparsity_tf2 import pruner
-from tensorflow_model_optimization.python.core.sparsity.keras import pruning_schedule
-from tensorflow_model_optimization.python.core.sparsity.keras import pruning_utils
+from tensorflow_model_optimization.python.core.sparsity_tf2 import schedule
 
 test = tf.test
 
-
 class ScheduleTest(test.TestCase, parameterized.TestCase):
+  """Test to verify custome Schedule behaviour for step parameters.
```
Comment (tomerk) [inline tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py:28]: ultra nit: typo in `custom`

File: tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py [hunk 3]
```diff
@@ -14,30 +14,361 @@
...
+    self.optimizer = tf.keras.optimizers.SGD()
+
+  def _construct_schedule(
+      self, schedule_type, begin_step, end_step, frequency=10, k=3, 
+      optimizer=None, use_default_optimizer=True, clipnorm=None, clipvalue=None):
+    # Uses default values for sparsity. We're only testing begin_step, end_step
+    # and frequency here as a basic requirement.
+    # Other variants of the Constant schedule may have extra parameters tested.
+    # use_default_optimizer is False if we are testing optimizer function
+    initial_drop_fraction = tf.Variable(0.5) # must be a Variable unlike the orthodox sparsity float
```
Comment (tomerk) [inline tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py]: Why must initial_drop_fraction be a variable? can we rewrite the tests to not use a variable here?

Or if the schedules require it to be a schedule, that's a little problematic (again due to serialization/deserialization challenges) & needs to be explicitly documented

File: tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py [hunk 4]
```diff
@@ -14,30 +14,361 @@
...
+          'schedule_type': 'exponential_decay'          
+      }, {
+          'testcase_name': 'LRSchedule',
+          'schedule_type': 'lr_decay'
+      })
+  def testPrunesOnlyInBeginEndStepRange(self, schedule_type):
+    decay = self._construct_schedule(schedule_type, 100, 200, 1)
+
+    # Before begin step
+    step_90 = tf.Variable(90)
```
Comment (tomerk) [inline tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py]: why are all these variables?

File: tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py (hunk 3 above)
Comment (tomerk) [reply tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py]: We talked about this offline: the tests should avoid using `tf.variable`s except in the specific case of a test checking dynamicism when the drop fraction variable is changed

File: tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py (hunk 4 above)
Comment (tomerk) [reply tensorflow_model_optimization/python/core/sparsity_tf2/schedule_test.py]: We talked about this offline: the tests should avoid using `tf.variable`s except in the specific case of a test checking dynamicism when the drop fraction variable is changed

**Review by tomerk — APPROVED** [review]
This looks good for now, thank you Winnie!
